src/Deserializer.cpp

```diff
@@ -16,15 +16,327 @@
+#include "Spot.h"
+#include "TimerManager.h"
 
-#include <climits>
+#include <vector>
 
 namespace dagon {
 
 ////////////////////////////////////////////////////////////
-// Implementation - Public
+// Implementation - Constructor & Destructor
 ////////////////////////////////////////////////////////////
+
+Deserializer::Deserializer(lua_State *L, SDL_RWops *rw) : _L(L),_rw(rw) {}
+
+Deserializer::~Deserializer() {
+  SDL_RWclose(_rw);
+}
+
+////////////////////////////////////////////////////////////
+// Implementation - Private
+////////////////////////////////////////////////////////////
+
+const char *Deserializer::readFunction(lua_State *L, void *data, size_t *size) {
+  LuaReaderData *readerData = reinterpret_cast<LuaReaderData*>(data);
+  void *p = const_cast<char*>(readerData->buf);
+  if (!SDL_RWread(readerData->rw, p, readerData->numBytes, 1)) {
+    readerData->readError = true;
+    *size = 0;
+    return nullptr;
+  }
+
+  *size = readerData->numBytes;
+  return readerData->buf;
+}
+
+////////////////////////////////////////////////////////////
+// Implementation - Gets
+////////////////////////////////////////////////////////////
+
+std::string Deserializer::version() {
+  return _dgVersion;
+}
+
+std::string Deserializer::preview() {
+  return _preview;
+}
+
+std::string Deserializer::roomName() {
+  return _roomName;
+}
+
+////////////////////////////////////////////////////////////
+// Implementation - State changes
+////////////////////////////////////////////////////////////
+
+bool Deserializer::readHeader() {
+  { // Read Dagon version string
+    uint8_t len;
+    if (!(len = SDL_ReadU8(_rw)))
+      return false;
+    std::vector<uint8_t> buf(len);
+    if (!SDL_RWread(_rw, buf.data(), buf.capacity(), 1))
+      return false;
+    _dgVersion = std::string(buf.begin(), buf.end());
+  }
+
+  { // Read preview text
+    uint8_t len;
+    if (!(len = SDL_ReadU8(_rw)))
+      return false;
+    std::vector<uint8_t> buf(len);
+    if (!SDL_RWread(_rw, buf.data(), buf.capacity(), 1))
+      return false;
+    _preview = std::string(buf.begin(), buf.end());
+  }
+
+  { // Read Room name
+    uint8_t len;
+    if (!(len = SDL_ReadU8(_rw)))
+      return false;
+    std::vector<uint8_t> buf(len);
+    if (!SDL_RWread(_rw, buf.data(), buf.capacity(), 1))
+      return false;
+    _roomName = std::string(buf.begin(), buf.end());
+  }
+
+  return true;
+}
+
+bool Deserializer::readScriptData() {
+  uint32_t numFields = SDL_ReadBE32(_rw);
+  for (uint32_t i = 0; i < numFields; i++) {
+    uint16_t lineLen = SDL_ReadBE16(_rw);
+    std::vector<uint8_t> lineBuf(lineLen);
+    if (!SDL_RWread(_rw, lineBuf.data(), lineBuf.capacity(), 1))
+      return false;
+
+    std::string line(lineBuf.begin(), lineBuf.end());
+    int errCode = luaL_dostring(_L, line.c_str());
+    if (errCode != 0) { // Perhaps one of the tables we are inserting into doesn't exist
+      // TODO: begin creating empty tables at the deepest level of the line
```

Comment:
My plans for this is to use a regex to decide on a substring to assign an empty table to. For instance suppose the line `dgPersistence["t"][1]=339` fails. Maybe table `t` doesn't exist yet? So by regex magic we create the substring `dgPersistence["t"]={}` and then we try the original line again. We will of course try more if needed. That is, go "another level up" and try to make an empty table there... I think this should work if we get the details right.